### ISSUE TITLE
@lru_cache interferes with @overload'ed definitions

### ISSUE THREAD

**Issue by bukzor** [microsoft/pyright#8692]
This is a simplified version of what's seen in stdlib `urllib.parse`.

Code sample in [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4DOTYGMkCgYC2YA9lEiAGYCuAdjscQDZq4VTEEhIRgw1whCJMiACCNCAGVMAGhAA5AIYEApgBMAKlTCMVcxWjQqyAfW5g9IYgDdjjYorW5nWRgbQhJOmEgBKKtCpGJAAKJVVNbV0AShRcEATwd2dcV3dPbz8AoKQAIQgkALDldS0dFVj4xLBk5wABGzsHJzUVChAqKGZMkM7GFBAMKGiQAFoAPgzGH39A4IGAOiXcBtsu5txW9r60Hr6BgCMCgJGJqZns4PzCtEXluuo6JAZmBcYoKhMsRSwEFRDzOoALwaD4VTZtDpdXbTUL7MQSaTDMaTLyw2Y5EAAH3OWTmeWOtyqCRg7RgaD4GEUdH%2BfTkQ0qiSZICgKiQnRouIxwRC0WJIBUzBUcWZiVZ7KgnLRF3x1yKfOcAA8QECod1YSEAETEBCKGCavnuYxIMw8f6KuTSvE5aJAA)

```python
# pyright:strict
import functools
from typing import AnyStr, NamedTuple, assert_type, overload


class SplitResult(NamedTuple):
    pass


class SplitResultBytes(NamedTuple):
    pass


@overload
def urlsplit(url: str) -> SplitResult: ...
@overload
def urlsplit(url: bytes) -> SplitResultBytes: ...
@functools.lru_cache(typed=True)
def urlsplit(url: AnyStr) -> SplitResult | SplitResultBytes:
    if isinstance(url, str):
        return SplitResult()
    else:
        return SplitResultBytes()


x = urlsplit("ohai")
assert_type(x, SplitResult)
```

As written, currently, this assert_type fails, but removing the lru_cache removes the error.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by bukzor** [comment]
I dug into this and found that it seems to be a bug in typeshed, rather -- lru_cache wrapper uses `...` as the paramspec which means pyright loses the overload information.

What did you find?


```python
from typing import Callable, ParamSpec, TypeVar, assert_type, overload

P = ParamSpec("P")
T = TypeVar("T")
def cached_broken(f: Callable[..., T]) -> Callable[..., T]: return f
def cached_worken(f: Callable[P, T]) -> Callable[P, T]: return f

@overload
def g(x: float) -> complex: ...
@overload
def g(x: str) -> None: ...
def g(x: float | str) -> complex | None:
    del x
    return None

cg = cached_broken(g)
assert_type(cg(1.2), complex)
assert_type(cg("1.2"), complex)

cg = cached_worken(g)
assert_type(cg(1.2), complex)
assert_type(cg("1.2"), None)
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.376.
